### DISCUSSION TITLE
Projectile Motion Exercise Not Working

### DISCUSSION THREAD

**Discussion by Nadinemm5** [cooperrc/computational-mechanics#162] [Q&A] [answered]
In the first module 3 reading, I can't seem to get the widget to advance the frames so I can collect the next data points. Im able to collect data points and I'm not getting any errors. The widget shows up and displays the numbers correctly, but it does not affect/refresh the image. Any ideas what the issue could be?
![image](https://user-images.githubusercontent.com/121512414/211210223-1e493fc9-c560-4d44-9e91-a300515457b4.png)

**Comment by cooperrc** [comment]
What happens when you click the image and run a cell with just the command, 

```python
coords
```

?

It should give you a list of [[x, y], [...]] locations where you clicked

**Reply by Nadinemm5** [reply]
I get coordinates just as I did in the previous sections. 
![image](https://user-images.githubusercontent.com/121512414/211390925-0b080fe3-72c7-46c6-857b-cc4e26205125.png)

**Reply by cooperrc** [reply]
ahh, so the problem is that the frames don't progress if you click the forward/backward buttons?

**Reply by Nadinemm5** [reply]
correct. The number will tick up/down with the arrows but the image does not refresh.

**Reply by cooperrc** [reply]
![image](https://user-images.githubusercontent.com/11371428/211392301-302ba951-8122-47fb-9a40-587db0577d2c.png)

**Reply by Nadinemm5** [reply]
![image](https://user-images.githubusercontent.com/121512414/211392838-9d613ed5-d8e5-484f-8e06-6ab4c7326da3.png)
this is what I get when I change the frame number

**Reply by cooperrc** [reply]
Usually, the issue has to do with conflicting variables in the matplotlib widget interfaces. 

- Dumb solution: (turn it off and back on again)
- click "Kernel" -> "restart"
- Go to the top of the notebook, then hold shift, then press enter until you get back to the **Capture mouse clicks on many frames with a widget** section and click the horizontal lines again then try to do the next part

**Reply by Nadinemm5** [reply]
I did restart the file/jupyter before, and restarted the kernel + re-ran the cells as you suggested, but nothing changed. (I run jupyter locally, maybe this is the issue? Something I need to install perhaps?)

**Reply by cooperrc** [reply]
> I did restart the file/jupyter before, and restarted the kernel + re-ran the cells as you suggested, but nothing changed. (I run jupyter locally, maybe this is the issue? Something I need to install perhaps?)

I'm running a test on my computer, you might need to replace `%matplotlib interactive` with `%matplotlib widget`

**Comment by cooperrc** [comment, marked as answer]
OK, I created a fresh conda environment and installed matplotlib, ipypl, imageio-ffmpeg, imageio, amd jupyterlab. Here's what I learned:
- I needed `%matplotlib notebook` to see the images and get interactive clicks
- I had to launch the notebook in Jupyter notebook, not JupyterLab
- Chrome worked better than Firefox
- I needed all the libraries to be up-to-date, with conda the command is `conda update --all`

**Reply by Nadinemm5** [reply]
Great! I was already doing points 1-3; I ran a new terminal and was able to check all my installations and update all; I think the issue was with ipympl. Oddly enough when I ran `pip install ipympl` It started to output `Requirement already satisfied` but at the end it said `Successfully installed ipympl-0.9.2` so I think it may have needed to be patched/updated.  
I restarted the kernel and it looks like everything is working now. I can finally see all the positions of the projectile !! Thanks for all your help.

**Comment by robert-avena** [comment]
I'm experiencing the same issue, I tried installing all those packages (I did not have ipympl) and doing ```conda update --all``` and re-running the import cells but I still cannot step through the frames. I'm also on a local install of jupyter. 

I've noticed in the terminal several "No such comm" warnings pop up every time I try to run that cell:

![image](https://user-images.githubusercontent.com/121516267/211404529-045d20fc-6428-4ef1-8dcd-2e541c2ccf90.png)

Not sure what these mean or if they could help solve the problem at all.

**Reply by cooperrc** [reply]
Awesome troubleshooting work, but from what I can find [`No such comm:`](https://github.com/jupyter-widgets/ipywidgets/issues/3454) is an unrelated warning that _shouldn't_ affect interactions.

**Reply by Nadinemm5** [reply]
For me the issue was fixed once I used `pip install ipympl` in a jupyter terminal, restarted the kernel, and shift+enter through the whole notebook until at the section once again to run any necessary cells. I skipped downloading the first example to save time.  I did not encounter any issues with this approach;  
Did you run the terminal inside of jupyter notebook? did you use `conda install` or `pip install`? maybe one works better than the other.
